universal-login-relayer/lib/http/routes/authorisation.ts

```diff
@@ -17,13 +22,30 @@ const getPending = (authorisationService : AuthorisationService) =>
     return responseOf({ response: result });
   };
 
-const denyRequest = (authorisationService : AuthorisationService) =>
-  async (data: {walletContractAddress: string, body: {key: string}}) => {
+const denyRequest = (authorisationService : AuthorisationService, provider: providers.Provider) =>
+  async (data: {walletContractAddress: string, body: {key: string, signature: utils.Signature}}) => {
+    const {walletContractAddress} = data;
+    const {key, signature} = data.body;
+
+    const cancelAuthorisationRequest: CancelAuthorisationRequest = {walletContractAddress, key};
+    const [isValid, computedAddress] = verifyCancelAuthorisationRequest(cancelAuthorisationRequest, signature, key);
+    if (!isValid) {
+      throw new InvalidSignature(`cancelAuthorisationRequest failed due to invalid signature`);
+    }
+
+    const contract = new ethers.Contract(walletContractAddress, WalletMasterWithRefund.interface, provider);
+    const flatSignature = ethers.utils.joinSignature(signature);
+    const payloadDigest = hashCancelAuthorisationRequest(cancelAuthorisationRequest);
+    const isCorrectAddress = await contract.isValidSignature(payloadDigest, flatSignature);
+    if (!isCorrectAddress) {
+      throw new InvalidAddress(computedAddress);
+    }
+
     const result = await authorisationService.removeRequest(data.walletContractAddress, data.body.key);
     return responseOf(result, 204);
```

Comment:
Everything except the last line should probably be extracted to `AuthorisationService`.